tools/extensions/phpstorm/Civi/PhpStorm/Api3Generator.php

```diff
@@ -0,0 +1,43 @@
+<?php
+
+namespace Civi\PhpStorm;
+
+use Civi\Core\Service\AutoService;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+/**
+ * @service civi.phpstorm.api3
+ */
+class Api3Generator extends AutoService implements EventSubscriberInterface {
+
+  public static function getSubscribedEvents() {
+    return [
+      'civi.phpstorm.flush' => 'generate',
+    ];
+  }
+
+  public function generate() {
+    /*
+     * FIXME: PHPSTORM_META doesn't seem to support compound dynamic arguments
+     * so even if you give it separate lists like
+     * ```
+     * expectedArguments(\civicrm_api4('Contact'), 1, 'a', 'b');
+     * expectedArguments(\civicrm_api4('Case'), 1, 'c', 'd');
+     * ```
+     * It doesn't differentiate them and always offers a,b,c,d for every entity.
+     * If they ever fix that upstream we could fetch a different list of actions per entity,
+     * but for now there's no point.
+     */
+
+    $entities = \civicrm_api3('entity', 'get', []);
+    $actions = ['create', 'delete', 'get', 'getactions', 'getcount', 'getfield', 'getfields', 'getlist', 'getoptions', 'getrefcount', 'getsingle', 'getunique', 'getvalue', 'replace', 'validate'];
+
+    $builder = new PhpStormMetadata('api3', __CLASS__);
+    $builder->registerArgumentsSet('api3Entities', ...$entities['values']);
+    $builder->registerArgumentsSet('api3Actions', ...$actions);
+    $builder->addExpectedArguments('\civicrm_api3()', 0, 'api3Entities');
+    $builder->addExpectedArguments('\civicrm_api3()', 1, 'api3Actions');
```

Comment:
Since you've defined a reusable set I guess it wouldn't hurt to add `civicrm_api` as well (it's basically v3-only).
```suggestion
    $builder->addExpectedArguments('\civicrm_api3()', 0, 'api3Entities');
    $builder->addExpectedArguments('\civicrm_api3()', 1, 'api3Actions');
    $builder->addExpectedArguments('\civicrm_api()', 0, 'api3Entities');
    $builder->addExpectedArguments('\civicrm_api()', 1, 'api3Actions');
```

I like how I've been busy ripping out APIv3 examples at the same time you were adding this. Yin an Yang 🙃

Reply:
@colemanw  lol. Yeah... I like to the think that the net-change is positive (total size of removed doc-examples exceeds extra size of these hints). 

It is easy to add `civicrm_api()`, and it's an acceptable cost. But I suppose #27180 gives a rationale for why we might not care to?

Reply:
Things that make you go "meh"